spec/Naming.md

```diff
@@ -0,0 +1,122 @@
+# Naming
+
+We define the unique name strategy per datasource to ensure it is followed uniformly independently from the type of job consuming and producing data.
+The namespace for a dataset is the unique name for its datasource.
+For jobs it is the scheduler.
+
+## Datasets:
+The namespace and name of a datasource can be combined to form a URI (scheme:[//authority]path)
+* Namespace = scheme:[//authority] (the datasource)
+* Name = path (the datasets)
+
+### Data warehouses/data bases.
+Datasets are called tables. Tables are organised in databases and schemas.
+#### Postgres:
+Datasource hierarchy:
+ * Host
+ * Port
+
+Naming hierarchy:
+ * Database
+ * Schema
+ * Table
+
+Identifier:
+ * Namespace: postgres://{host}:{port} of the service instance.
```

Comment:
JDBC connection URIs include the database name. Should the clients be responsible for parsing the connection URI to extract the database name and only report the host/port? The connection URI is going to vary by driver (e.g., oracle, postgres, and sqlite all have different uri formats).

Should we consider leaving the database name as part of the namespace and only including the schema/table in the name? That'll make it easier for clients to capture that information without parsing the different URIs.

Reply:
That makes sense. 
This spec is trying to stay close to the current Marquez implementation (dataset name = {db}.{schema}.{table})
So it's a bit of a trade off.
I'm assuming we would want to change the other db types as well to have the names consistently in 2 levels?

Reply:
In my mind, there's a tradeoff between consistency, existing conventions, and ease of implementation. Personally, I'd simply prefer not to parse JDBC connection URIs, if I can avoid it (have you seen Oracle connection URIs? 😬). Implementation details do seem to vary among databases. E.g., a Snowflake connection doesn't require the database in the connection URI- https://docs.snowflake.com/en/user-guide/jdbc-configure.html . The spark connector seems to use the same connection URI w/o database in the URI. BigQuery has no connection URI, so implementation is completely flexible. Hive connection URIs don't specify the database. Queries can use fully qualified table names or use the `USE <db>` statement.

I don't know that there has to be 100% consistency between each of these, but if consistency is important, I think it would be easier to move the `database` portion into the namespace than to extract it from the connection URI. If consistency _and_ existing convention is important, we'll need parsers for each flavor of JDBC connection URIs to extract database and form the FQ dataset name.

Reply:
A couple of points:
- We should consistently build those dataset names for all SQL data{base,warehouse}. I'd be happy with standardizing on `{schema}.{table}`. But it should be we should follow the same convention for bigquery and Snowflake for example.  
- We cannot get away with not looking at the connection URL. The namespace must be canonical as well. The fact that some parameters are added to a connection URL or that the database is optional should not lead to different names. Those names need to be canonical, if there's a default db name, then we need to add it explicitly in the namespace here.